File: hard/224. Basic Calculator/224. Basic Calculator.py
from typing import List


class Solution:
    def calculate(self, s: str) -> int:
        stack: List[int] = []
        result = 0
        prev_number = 0

        sign = 1

        def update_result() -> int:
            return result + sign * prev_number

        for element in s:
            if element.isdigit():
                prev_number = prev_number * 10 + int(element)

            elif element in "+-":
                result = update_result()
                prev_number = 0
                sign = 1 if element == "+" else -1

            elif element == "(":
                stack.append(result)
                stack.append(sign)
                result = 0
                sign = 1

            elif element == ")":
                result = update_result()
                prev_number = 0
                result *= stack.pop()
                result += stack.pop()

        result = update_result()
        return result


solution = Solution()
print(solution.calculate(s="1 + 1"))  # 2
print(" ")
print(solution.calculate(s=" 2-1 + 2 "))  # 3
print(" ")
print(solution.calculate(s="(1+(4+5+2)-3)+(6+8)"))  # 23
print(" ")
print(solution.calculate(s="2147483647"))
print(" ")
print(solution.calculate(s="1-(     -2)"))  # 3
print(" ")
print(solution.calculate(s="-2+ 1"))  # -1
print(" ")
print(solution.calculate(s="- (3 + (4 + 5))"))  # -12

File: hard/224. Basic Calculator/224. Basic Calculator.go
package main

import (
	"unicode"
)

func calculate(s string) int {
	stack := []int{}
	prevNumber := 0
	result := 0
	sign := 1

	for _, element := range s {
		if unicode.IsDigit(element) {
			prevNumber = prevNumber*10 + int(element-'0')
		} else if element == '+' || element == '-' {
			result += result + sign*prevNumber
			prevNumber = 0

			if element == '+' {
				sign = 1
			} else {
				sign = -1
			}
		} else if element == '(' {
			stack = append(stack, result)
			stack = append(stack, sign)
			result = 0
			sign = 1
		} else if element == ')' {
			result += result + sign*prevNumber
			prevNumber = 0

			result *= stack[len(stack)-1]
			stack = stack[:len(stack)-1]

			result += stack[len(stack)-1]
			stack = stack[:len(stack)-1]
		}
	}

	result += result + sign*prevNumber
	return result
}
